src-ext/TypeScript/ggvega/src/itmEncodingObject.ts

```diff
@@ -0,0 +1,219 @@
+import {encodingByAes} from './encodingByAes';
+import {GGEncodingKey} from './encodingNameByGeom';
+import {encodingValueColor, encodingValueShape, encodingValueSize} from './encodingValue';
+
+/**
+ * Create an intermediate `encoding` object using a `mapping` object
+ *
+ * @remark
+ * This function creates an intermdiate `encoding` object using a
+ * ggspec `mapping` object and a ggspec `Metadata` object.
+ *
+ * Currently, a ggspec `mapping` is an object with a single key, `field`.
+ * Its value is a string; that string matches to a variable-name
+ * in the data.
+ *
+ * In the near future, a ggspec `mapping` may instead contain a different
+ * single key, `stat`, with a string that matches an stat (aggregation)
+ * operation, like `"count"`, or `"sum"`.
+ *
+ * This is more of a note to myself, we are starting to see the patterns
+ * of how ggplot2 mappings, aes_params, and stat default_aes are related.
+ * I suspect this should be addressed in ggschema.
+ * See <https://github.com/vegawidget/ggvega/issues/44>
+ *
+ * **Called by**
+ *
+ * @see {@link itmLayer} to create an intermediate layer
+ *
+ * **Calls**
+ *
+ * @see {@link encodingByAes} to create an empty encoding
+ * @see {@link fieldName} to handle dots, ".", in field names
+ *
+ * @param ggMappingObject - `GG.Mapping` maps data varaibles to aesthetics
+ * @param ggMetadataObject - `GG.Metadata` contains the metadata for the data
+ *   associated to this layer
+ *
+ * @returns `ItmEncodingObject`
+ */
+export function itmEncodingObjectByMappingObject(
+  ggMappingObject: GG.Mapping,
+  ggMetadataObject: GG.Metadata
+): ItmEncodingObject {
+  // translate
+
+  // create empty itmEncodingObject
+  const itmEncodingObject: ItmEncodingObject = {};
+
+  // TODO: if the type is `ordinal`, and we have level,
+  // we should set the scale domain according to the levels.
+  //
+  // larger TODO: determine and document the relationship between
+  // ggplot    - factor, ordered-factor, levels
+  // Vega-Lite - nominal, ordinal, scale-domain
+
+  // for each mapping in gsMappingObject:
+  //   - extract the information from the mapping object, metadata
+  //   - create Encoding
+  //   - populate Encoding
+  //   - put Encoding into itmEncodingObject
+  for (const aesName in ggMappingObject) {
+    if (Object.prototype.hasOwnProperty.call(ggMappingObject, aesName)) {
+      // do we have a type/class for `mapping`?
+
+      // extract information from mapping object, metatdata
+      const mapping: GG.Encoding = ggMappingObject[aesName as GGEncodingKey] as GG.Encoding;
+
+      // TODO: we need to handle the situation where the mapping is a
+      // `stat` instead of a `field`
+
+      // NOTE @wenyu: Define `type` before we change the value of `field`
+      const type: VL.StandardType = ggMetadataObject[mapping.field].type;
+      const field: string = fieldName(mapping.field);
+
+      // create Encoding
+      const encoding: VLMapping = encodingByAes(aesName);
```

Comment:
I don't know what we can do about it now, but I want to draw some more attention to the future situation where an encoding could be a `stat`, I suspect we will have to specify a type that is more flexible than `VLMapping`.